File: resources/getOaLocation.js
require('dotenv').config()
const Fuse = require('fuse.js')

const { postLocation, retrieveAccessToken, deleteLocation, getLocations } = require("./manualHttpRequests");

// const AGENDA_UID = process.env.AGENDA_UID
const TBD_LOCATION_UID = process.env.TBD_LOCATION_UID
// const publicKey = process.env.OA_PUBLIC_KEY;
const secretKey = process.env.OA_SECRET_KEY;

/**
 * Search a given adress in OA locations DB (using name & adress) with FuzeJS
 * 
 * @param {string} eventlocation - The event location to search for e.g. "MJC Le Sterenn, Tregunc"
 * @param {array} locationsData - The list of all locations fetched from OpenAgenda.
 * @return {array} An array of best matching locations with {OALocationUid,NameAdress, roundedScore} sorted by best score
 */
const searchOALocation = (eventlocation, locationsData) => {
    // TODO: (?) Tweak FuzeJS to return nothing when score is not high enough
    const options = {
        minScore: 0.4,
        scorePrecision: 4,
        threshold: 0.5,
    }

    const fuzeOptions = {
        includeScore: true,
        ignoreLocation: true,
        keys: [
            { name: 'name', weight: 1.5 },
            { name: 'address', weight: 0.3 },
        ],
    }
    const fuse = new Fuse(locationsData, fuzeOptions, debug=false)
    const results = fuse.search(eventlocation)
    if (results.length > 0) {
        if (debug) console.log("Best result: ", results[0].item, "score: ", results[0].score)
        const precision = 10 ** options.scorePrecision
        // FIXME: Understand what is the "best" score in order to sort results
        const SelectedResult = results
            .filter(entry => entry.score < options.minScore)
            .map(entry => ({
                OALocationUid: entry.item.uid,
                NameAdress: `${entry.item.name} - ${entry.item.address}`,
                roundedScore: Math.round(entry.score * precision) / 10 ** precision,
                score: entry.score
            }))[0];
        return SelectedResult ?? null
    }
    return null
};

/**
 * Get a good matching location in OpenAgenda or empty object.
 *
 * @param {string} inputLocations  - An adress or place name.
 * @return {Promise<Array>}     - Object with found OA location {OALocationuid, NameAdress, roundedScore}.
 *                                or {} if no match found.
 *                           
 */
const matchOaLocations = async (inputLocation, accessToken, debug=false) => {
        if (debug) console.log("Searching match for ", inputLocation);

    const OALocations = await getLocations(accessToken) ?? [];
        if (!inputLocation || OALocations?.length == 0) {
            console.log("[ERROR] inputLocation is null or OA Locations list is empty ")
                return {}
        }
        if (debug) console.log("Number of OAlocations: ", OALocations.length);
    const searchResultOA = searchOALocation(inputLocation, OALocations)
        if ( ! searchResultOA?.NameAdress) {
            if (debug) console.log("[ERROR] No match found in existing OA locations ")
                return {}
        }
    
        if (debug) console.log("[SUCCESS] Match with '", searchResultOA.NameAdress,"'")
    return searchResultOA

};

const getCorrespondingOaLocation = async (inputLocation, debug=false) => {
    const accessToken = await retrieveAccessToken(secretKey)

    // try to find an existing OALocation
    // CAVEAT : Fuze is lost with long string search. Keeping only first 15 characters
    const OALocation = await matchOaLocations(inputLocation.substring(0, 20), accessToken, debug)

    if (OALocation?.NameAdress && OALocation?.OALocationUid) {
        // Case 1: Best case scenario where match is found
        // FIXME: What if we have a false positive e.g. inputLocation = "Mediathèque Elliant" and OaLocation = "Médiathèque Concarneau"
        if (debug) console.log("--> Returning existing OALcocation: ", OALocation.OALocationUid)
        return OALocation.OALocationUid
    } else if (inputLocation) {
        // Case 2 :No match found so we try to create the event with Open Agenda GeoCoding API
        // CAVEAT:  Hard to distinguih location name from postal address. We sends all location info to AO API.
        //          NB: Spliting name and address is specific to Google Agenda syntax (cannot be used as is for other sources e.g. when scraping)
        const response = await postLocation(accessToken, inputLocation)
        if (!response?.location?.uid) {
            // Case 2.1 : Nothing created
            if (debug) console.log("--> Returning location  'To be defined' (Could not create location on OpenAgenda)")
            return TBD_LOCATION_UID
        }
        const newOALocation = response.location
        console.log("New OAlocation created (to be validated): ", newOALocation.uid)
    
        // check if created location is more or less in Breizh
        const lat = newOALocation.latitude
        const long = newOALocation.longitude
        if (newOALocation?.uid
            && parseFloat(lat) < 49 && parseFloat(lat) > 47
            && parseFloat(long) < -1 && parseFloat(long) > -5.5
        ) {
            // Case 2.2 : Valid location in Breizh
            if (debug) console.log("--> Returning a valid new OA location: ", newOALocation.name, newOALocation.address)
            return newOALocation.uid    
        } else {
            // Case 2.3 : Invalid location, not in Breizh. Delete it.
            const data = await deleteLocation(accessToken, newOALocation.uid)
            if (debug) console.log("Location not in Breizh. Newly created OA Location was deleted :", newOALocation.uid, data.success ? "success" : "failed" + data)
            if (debug) console.log("--> Returning 'To be defined' location")
            return TBD_LOCATION_UID
        }
    } else {
        return TBD_LOCATION_UID
    }

}

module.exports = { getCorrespondingOaLocation }

/**
 * Test getCorrespondingOaLocation() with different exemples             
 */

// const LocationsExemples = [
//     { inputLocation: 'MJC Tregunc Le Sterenn, Rue Jacques Prévert, 29910 Trégunc, France'  },
//     { inputLocation: 'MJC Tregunc Le Sterenn, Rue Jacques Prévert'},
//     { inputLocation: 'MJC Tregunc Le Sterenn'},
//     { inputLocation: 'Le Sterenn'},
//     { inputLocation: 'MJC Tregunc'},
    // { inputLocation: 'Explore'  },
    // { inputLocation: "Bar de Test, 1 Pl. de l'Église, 29100 Pouldergat"  },
    // { inputLocation: 'qsdfg'  },
    // { inputLocation: '30 Rue Edgar Degas, 72000 Le Mans'  },
    // { inputLocation: '11 Lieu-dit, Quilinen, 29510 Landrévarzec'},
    // { inputLocation: 'Kerminy, Rosporden'},
    // { inputLocation: 'CONCARNEAU, Carré des larrons'}


// async function testLocations( locationArray) {
//     for (const loc of locationArray) {
//         const testUid = await getCorrespondingOaLocation(loc.inputLocation, debug=true);
//         console.log("matching --> OA Uid=", testUid);
//         console.log("--------------");
//     }
// };

// testLocations(LocationsExemples);

File: resources/python/getOaLocation.py
"""_summary_
Functions to manage locations with OA API
"""

import sys
import os
import git

# Ajoute le dossier "ressources" au sys.path
git_root = git.Repo(search_parent_directories=True).working_tree_dir
sys.path.insert(0,   os.path.abspath(  os.path.join(  git_root,'resources/python' ) ) )

from utils import *
from scraping_utils import *
from HttpRequests import post_location, retrieve_access_token, delete_location, get_locations

from thefuzz import fuzz
from thefuzz import process
import re

# Environment variables
TBD_LOCATION_UID = os.getenv("TBD_LOCATION_UID")


def get_or_create_oa_location(searched_location:str, access_token: str, debug:bool=False)->str:
    """
    Tries to find a matching OpenAgenda location for the given searched location.
    Returns an OALocation UID (found, created or default one.)
    """
    print("- searching location for : '"+ searched_location +"'")
    if (searched_location == None ) or (searched_location.lower() in ( "" , "none", "null")):
        print("InputLocation is null or empty. Returning default Location")
        return TBD_LOCATION_UID
    allOaLocations = get_locations(access_token)
    if not searched_location or not allOaLocations:
        print("[ERROR] inputLocation is null or OA Locations list is empty")
        return None
    
    # 0) Use optimized searched, removing false positives and misleadings patterns
    locationPatterneRemoved=re.sub(r'\b(?:concarneau|(?:29|56)\d{3}|officiel|quimperl(?:e|é)|france |spectacles)\b',
                                            '',
                                            searched_location,
                                            flags=re.IGNORECASE
                                            )
    locationPatternToSpace=re.sub(r'[-,():]',
                                            ' ',
                                            locationPatterneRemoved,
                                            flags=re.IGNORECASE
                                            )
    optimized_searched_location = locationPatternToSpace
    if debug : print(" (optimized name for better matching:  '"+ optimized_searched_location +"')")
    # 1) Try to find an existing OALocation
    OaLocationsIndex = {}
    for location in allOaLocations:
        OaLocationsIndex[location["uid"]]=location['name'] + " " + location['address']
    # returns a list of tuples (name adress , score , OAuid)
    results = process.extract(optimized_searched_location, OaLocationsIndex, scorer=fuzz.token_set_ratio) or []
    if results[0] and results[0][1] > 85:  # Best matching score >85
        print(f"- 🎯 Location found in OA: {results[0] }")
        return results[0][2]

    # 2) Try to create an OALocation
    response = post_location(access_token, searched_location, searched_location)
    if not response or not response.get('location', {}).get('uid'):
        print("-> ❔ Returning location 'To be defined' (Could not create location on OpenAgenda)")
        return TBD_LOCATION_UID

    # Stay in rectangle covering Breizh
    new_oa_location = response['location']
    lat = new_oa_location['latitude']
    long = new_oa_location['longitude']
    if ( new_oa_location.get('uid')        
        and 47 < float(lat) < 49
        and -5.5 < float(long) < -1 
        ):
        print(f"-> '\U0001f195' New OA location created : {new_oa_location['name']}, {new_oa_location['address']}, {"https://openagenda.com/kerlandrier/admin/locations/" + str(new_oa_location['uid'])}")
        return new_oa_location['uid']
    else:
        delete_location(access_token, new_oa_location['uid'])
        print(f"-> ❔ Location not in Breizh: returning 'To be defined' location")
        return TBD_LOCATION_UID

def get_locations_list(searched_location:str, access_token: str)->list:
    """
    Return list of matching locations in format ('NAME, ADRESS, SCORE [0-100], UID)
    """
    allOaLocations = get_locations(access_token)
    if not searched_location or not allOaLocations:
        print("[ERROR] inputLocation is null or OA Locations list is empty")
        return None
    
    # 1) Try to find an existing OALocation
    OaLocationsIndex = {}
    for location in allOaLocations:
        OaLocationsIndex[location["uid"]]=location['name'] + " " + location['address']
    # returns a list of tuples (name adress , score , OAuid)
    results = process.extract(searched_location, OaLocationsIndex, scorer=fuzz.token_set_ratio) or []
    
    
    if len(results): 
        if results[0][1] >= 80:  # Best matching score >85
            return results[0]
        # good_results = [item for item in results if item["score"] > 70]
        return results[:5]
    else:
        return None

#####################
## Test cases:
#####################

locations_examples = [
    {"input_location": 'MJC Tregunc Le Sterenn, Rue Jacques Prévert, 29910 Trégunc, France', "expectedUID": 89326663},
    {"input_location": 'Explore', "expectedUID": 5705265},
    {"input_location": "Bar de Test, 1 Pl. de l'Église, 29100 Pouldergat", "expectedUID": TBD_LOCATION_UID}, # Lieu inexistant mais ville existante
    {"input_location": 'qsdfg', "expectedUID": TBD_LOCATION_UID}, # Texte aléatoire
    {"input_location": '30 Rue Edgar Degas, 72000 Le Mans', "expectedUID": TBD_LOCATION_UID}, # HOrs Bretagne
    {"input_location": '11 Lieu-dit Quilinen 29510 Landrévarzec'}, # Non répertorié sur OA
    {"input_location": 'La Loco', "expectedUID": 34261153},
    {"input_location": 'Boulevard de la Gare, Quimperlé', "expectedUID": 34261153},
    {"input_location": 'La Caserne Concarneau ', "expectedUID": 9308588},
    {"input_location": '1 avenue Docteur NICOLAS, Concarneau'},
    {"input_location": 'Intermarché Concarneau (Route de Trégunc, Concarneau)', "expectedUID": 75052765},
    {"input_location": 'Brasserie Tri Martolod-Officiel', "expectedUID": 16309876},
    {"input_location": 'Rue de Colguen, 29900 Concarneau', "expectedUID": 24412066},
    {"input_location": 'Boulevard de la Gare, 29300 Quimperlé, France', "expectedUID": 34261153},
    
]

def test_locations(location_array):
    SECRET_KEY = os.getenv("OA_SECRET_KEY")
    access_token = retrieve_access_token(SECRET_KEY)
    allLocationsOA = get_locations(access_token)
    allLocationsOA_by_uid = {item['uid']: item for item in allLocationsOA}
    nbr_locations= len(allLocationsOA)
    print(f"Number of locations: {nbr_locations}")
    for loc in location_array:
        uid = get_or_create_oa_location( loc.get("input_location"),access_token, debug=True)
        if loc.get("expectedUID") and (loc.get("expectedUID") == uid): 
            print(" - ✅ Match with Expected location.")
        elif loc.get("expectedUID"): 
            print(" - ❌ Does not match with Expected location: '", allLocationsOA_by_uid.get (loc.get("expectedUID")).get("name"),"'")
        print("--------------\n")

if __name__ == "__main__":
    test_locations(locations_examples)
